### PR TITLE
8297582: C2: very slow compilation due to type system verification code

### PR DESCRIPTION
**Pull request by rwestrel**
The problem here is that for arrays, verification code computes a
number of meet operations that grows exponentially with the number of
dimensions while the number of unique meet operations that need to be
computed is a linear function of the number of dimensions:

```
// With verification code, the meet of A and B causes the computation of:
// 1- meet(A, B)
// 2- meet(B, A)
// 3- meet(dual(meet(A, B)), dual(A))
// 4- meet(dual(meet(A, B)), dual(B))
// 5- meet(dual(A), dual(B))
// 6- meet(dual(B), dual(A))
// 7- meet(dual(meet(dual(A), dual(B))), A)
// 8- meet(dual(meet(dual(A), dual(B))), B)
//
// In addition the meet of A[] and B[] requires the computation of the meet of A and B.
//
// The meet of A[] and B[] triggers the computation of:
// 1- meet(A[], B[][)
//   1.1- meet(A, B)
//   1.2- meet(B, A)
//   1.3- meet(dual(meet(A, B)), dual(A))
//   1.4- meet(dual(meet(A, B)), dual(B))
//   1.5- meet(dual(A), dual(B))
//   1.6- meet(dual(B), dual(A))
//   1.7- meet(dual(meet(dual(A), dual(B))), A)
//   1.8- meet(dual(meet(dual(A), dual(B))), B)
// 2- meet(B[], A[])
//   2.1- meet(B, A) = 1.2
//   2.2- meet(A, B) = 1.1
//   2.3- meet(dual(meet(B, A)), dual(B)) = 1.4
//   2.4- meet(dual(meet(B, A)), dual(A)) = 1.3
//   2.5- meet(dual(B), dual(A)) = 1.6
//   2.6- meet(dual(A), dual(B)) = 1.5
//   2.7- meet(dual(meet(dual(B), dual(A))), B) = 1.8
//   2.8- meet(dual(meet(dual(B), dual(A))), B) = 1.7
// etc.
```

There are a lot of redundant computations being performed. The fix I
propose is simply to cache the result of meet computations. So whene
the type system code is called to compute, for instance, the meet of
A[][] and B[][], the cache starts empty. Then as the meet computations
proceed, the cache is filled with meet result for meet of A[] and B[],
meet of A and B etc. Once the type system code returns with the result
for A[][] and B[][], the cache is cleared.

With this, the test case I added goes from "never seem to ever finish"
to "complete in no time".

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297582](https://bugs.openjdk.org/browse/JDK-8297582): C2: very slow compilation due to type system verification code

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [b28fde54](https://git.openjdk.org/jdk/pull/11673/files/b28fde544690f73e5933e11976e7000f97abb68a)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [b28fde54](https://git.openjdk.org/jdk/pull/11673/files/b28fde544690f73e5933e11976e7000f97abb68a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11673/head:pull/11673` \
`$ git checkout pull/11673`

Update a local copy of the PR: \
`$ git checkout pull/11673` \
`$ git pull https://git.openjdk.org/jdk pull/11673/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11673`

View PR using the GUI difftool: \
`$ git pr show -t 11673`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11673.diff">https://git.openjdk.org/jdk/pull/11673.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/compile.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/compile.cpp:654]: I assume it is typo: `_verif_`. Please use `_verify_`.

File: src/hotspot/share/opto/compile.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/compile.cpp:1084]: Please use `_type_verify_depth` to show that it is related to the cache.

File: src/hotspot/share/opto/type.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/type.cpp:845]: Please change to `VerifyMeetMark`.

File: src/hotspot/share/opto/type.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/type.cpp:918]: Can `VerifyMeetResult` class be local to `Type` class instead of `Compile` since it is used only locally here and you reset cache each time anyway? (`Compile` is become very big with verification code from all parts of compiler).

I thought you build the cache for duration of compilation then it may make sense to keep it in `Compile`. But you are resetting it after each `meet` so you don't need to have it in `Compile`.

Comment (rwestrel) [reply src/hotspot/share/opto/type.cpp:918]: What I could do is have a pointer to the cache as a field in Compile but move the new class declaration out of compile.hpp in type.[ch]pp? Does that sound ok to you?

Comment (vnkozlov) [reply src/hotspot/share/opto/type.cpp:918]: Yes, have only pointer to cache in Compile class is fine.

**Comment by rwestrel** [comment]
@vnkozlov I think I addressed all your comments.

**Review by vnkozlov — COMMENTED** [review]
Nice. I will submit testing.

You need second review.

File: src/hotspot/share/opto/type.cpp
Comment (iwanowww) [inline src/hotspot/share/opto/type.cpp:989]: As a style nit, you could access `VerifyMeetResult*` through `VerifyMeetMark` or even encapsulate it there.

**Review by iwanowww — APPROVED** [review]
Looks good.

File: src/hotspot/share/opto/type.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/type.cpp:981]: I missed this. `verif(C)` -> `verify()`

File: src/hotspot/share/opto/type.cpp
Comment (vnkozlov) [reply src/hotspot/share/opto/type.cpp:989]: I assume you suggested to hide access  to cache (`add()`, `meet()`) in `VerifyMeetMark`. I agree. `VerifyMeetMark` can be renamed `VerifyMeet` in such case.

File: src/hotspot/share/opto/type.hpp
Comment (iwanowww) [inline src/hotspot/share/opto/type.hpp:188]: Just noticed that `PRODUCT_RETURN` breaks optimized build (`Type::assert_type_verify_empty()` is under `ASSERT`).

**Review by iwanowww — CHANGES_REQUESTED** [review]

**Comment by rwestrel** [comment]
Thanks for the suggestions @vnkozlov @iwanowww. The new commit should address them.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by rwestrel** [comment]
FTR, merged with upstream now that interface changes are in.

**Comment by rwestrel** [comment]
FTR, solved merge conflicts

**Pull request closed without merging**
